### PR TITLE
Custom connection pool to MultiClusterPooledConnectionProvider

### PR DESCRIPTION
**Pull request by thachlp**
Issue: https://github.com/redis/jedis/issues/3753

### CODE REVIEW

File: src/main/java/redis/clients/jedis/MultiClusterClientConfig.java [hunk 1]
```diff
@@ -177,10 +177,19 @@ public static class ClusterConfig {
         private int priority;
         private HostAndPort hostAndPort;
         private JedisClientConfig clientConfig;
+        private ConnectionPoolConfig connectionPoolConfig;
 
         public ClusterConfig(HostAndPort hostAndPort, JedisClientConfig clientConfig) {
             this.hostAndPort = hostAndPort;
             this.clientConfig = clientConfig;
+            connectionPoolConfig = new ConnectionPoolConfig();
```
Comment (thachlp) [inline src/main/java/redis/clients/jedis/MultiClusterClientConfig.java:185]: Should we add a default ConnectionPoolConfig?

**Review by sazzad16 — CHANGES_REQUESTED** [review]
Apologies. I reviewed this some time ago. Just now I've realized I never _submitted_ it.

File: src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java [hunk 2]
```diff
@@ -132,7 +133,7 @@ public MultiClusterPooledConnectionProvider(MultiClusterClientConfig multiCluste
 
             multiClusterMap.put(config.getPriority(),
                     new Cluster(new ConnectionPool(config.getHostAndPort(),
-                            config.getJedisClientConfig()), retry, circuitBreaker));
+                            config.getJedisClientConfig(), poolConfig), retry, circuitBreaker));
```
[the remark this answers is not in the document]
Comment (thachlp) [reply src/main/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProvider.java]: Nice, I updated

File: src/main/java/redis/clients/jedis/MultiClusterClientConfig.java [hunk 3]
```diff
@@ -177,12 +178,20 @@ public static class ClusterConfig {
...
         private JedisClientConfig clientConfig;
+        private GenericObjectPoolConfig<Connection> connectionPoolConfig;
 
         public ClusterConfig(HostAndPort hostAndPort, JedisClientConfig clientConfig) {
             this.hostAndPort = hostAndPort;
             this.clientConfig = clientConfig;
         }
 
+        public ClusterConfig(HostAndPort hostAndPort, JedisClientConfig clientConfig,
+                             ConnectionPoolConfig connectionPoolConfig) {
```
Comment (sazzad16) [inline src/main/java/redis/clients/jedis/MultiClusterClientConfig.java:189]:
```suggestion
                             GenericObjectPoolConfig<Connection> connectionPoolConfig) {
```

File: src/test/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProviderTest.java [hunk 4]
```diff
@@ -131,4 +131,26 @@ public void testSetActiveMultiClusterIndexOutOfRange() {
...
+            MultiClusterPooledConnectionProvider.Cluster activeCluster = provider.getCluster();
+            ConnectionPool connectionPool = activeCluster.getConnectionPool();
+            assertEquals(8, connectionPool.getMaxTotal());
+            assertEquals(4, connectionPool.getMaxIdle());
+            assertEquals(1, connectionPool.getMinIdle());
+        } finally {
+            if (provider != null) {
+                provider.close();
+            }
+        }
```
Comment (sazzad16) [inline src/test/java/redis/clients/jedis/providers/MultiClusterPooledConnectionProviderTest.java:154]:
```suggestion
        try (MultiClusterPooledConnectionProvider customProvider = new MultiClusterPooledConnectionProvider(new MultiClusterClientConfig.Builder(clusterConfigs).build())) {
            MultiClusterPooledConnectionProvider.Cluster activeCluster = customProvider.getCluster();
            ConnectionPool connectionPool = activeCluster.getConnectionPool();
            assertEquals(8, connectionPool.getMaxTotal());
            assertEquals(4, connectionPool.getMaxIdle());
            assertEquals(1, connectionPool.getMinIdle());
        }
```

**Review by sazzad16 — CHANGES_REQUESTED** [review]

**Review by sazzad16 — APPROVED** [review]

**Comment by sazzad16** [comment]
Merged.

Thanks @thachlp for your contribution.

**Comment by thachlp** [comment]
Thanks @sazzad16

**Pull request merged**